apps/aecore/src/aec_peer_connection.erl

```diff
@@ -1234,3 +1245,51 @@ get_micro_block_txs([TxHash | TxHashes], State, Acc) ->
         none         -> get_micro_block_txs(TxHashes, some, [TxHash | Acc]);
         {value, STx} -> get_micro_block_txs(TxHashes, State, [STx | Acc])
     end.
+
+pre_assembly_check(MicroHeader) ->
+    {ok, Hash} = aec_headers:hash_header(MicroHeader),
+    case aec_db:has_block(Hash) of
+        true -> known;
+        false ->
+            Validators = [fun validate_micro/1,
+                          fun validate_connected_to_chain/1,
+                          fun validate_delta_height/1,
+                          fun validate_micro_signature/1],
+            aeu_validation:run(Validators, [MicroHeader])
+    end.
+
+validate_connected_to_chain(MicroHeader) ->
+    case aec_chain_state:hash_is_connected_to_genesis(
+            aec_headers:prev_hash(MicroHeader)) of
+        true -> ok;
+        false -> {error, orphan_blocks_not_allowed}
+    end.
+
+validate_delta_height(MicroHeader) ->
+    Height = aec_headers:height(MicroHeader),
+    case aec_chain:top_header() of
+        undefined -> ok;
+        TopHeader ->
+            MaxDelta = aec_chain_state:gossip_allowed_height_from_top(),
+            case Height >= aec_headers:height(TopHeader) - MaxDelta of
+                true  -> ok;
+                false -> {error, too_far_below_top}
+            end
+    end.
+
+validate_micro(MicroHeader) ->
+    aec_headers:validate_micro_block_header(MicroHeader).
+
+validate_micro_signature(MicroHeader) ->
+    case aec_db:find_header(aec_headers:prev_key_hash(MicroHeader)) of
+        none -> {error, signer_not_found}; %% This is virtually impossible!
```

Comment:
It is not impossible since you didn't check that the prev_key_hash is pointing correctly. It is impossible if you do check that.

Reply:
Fixed
|
|
v